components/brave_rewards/browser/rewards_service_impl.cc

```diff
@@ -1112,28 +1113,67 @@ void RewardsServiceImpl::OnURLLoaderComplete(
-    return data_decoder::JsonSanitizer::Sanitize(
-        *response_body,
-        base::BindOnce(
-            [](ledger::client::LoadURLCallback callback,
-               ledger::mojom::UrlResponse response,
-               data_decoder::JsonSanitizer::Result result) {
-              if (result.value) {
-                response.body = std::move(*result.value);
-              } else {
-                response.body = {};
-                VLOG(0) << "Response sanitization error: "
-                        << (result.error ? *result.error : "unknown");
-              }
-
-              std::move(callback).Run(response);
-            },
-            std::move(callback), std::move(response)));
+    return OnURLLoaderCompleteSanitizeJson(
+        std::move(callback), std::move(response), std::move(response_body));
   }
 
   std::move(callback).Run(response);
 }
 
+namespace {
+
+void JsonSanitizeCallback(
+    ledger::client::LoadURLCallback callback,
+    ledger::mojom::UrlResponse response,
+    const absl::optional<scoped_refptr<base::SequencedTaskRunner>>&
+        post_response_runner,
+    data_decoder::JsonSanitizer::Result result) {
+  if (result.value) {
+    response.body = std::move(*result.value);
+  } else {
+    response.body = {};
+    VLOG(0) << "Response sanitization error: "
+            << (result.error ? *result.error : "unknown");
+  }
+
+  if (post_response_runner) {
+    (*post_response_runner)
+        ->PostTask(FROM_HERE,
+                   base::BindOnce(std::move(callback), std::move(response)));
+  } else {
+    std::move(callback).Run(response);
+  }
+}
+
+}  // namespace
+
+void RewardsServiceImpl::OnURLLoaderCompleteSanitizeJson(
+    ledger::client::LoadURLCallback callback,
+    ledger::mojom::UrlResponse response,
+    std::unique_ptr<std::string> response_body) {
+  // On Android |RewardsServiceImpl::OnURLLoaderComplete| is executed in UI
+  // thread. We have evidences of ANRs happened at |JsonSanitizer::Sanitize|.
+  // By this reason sanitizing on Android is moved to non-UI thread, callback
+  // is still invoked in UI thread.
+  // Desktop in fact doesn't need this, decision is to have the same code for
+  // different platforms.
+  json_sanitizer_task_runner_->PostTask(
+      FROM_HERE,
+      base::BindOnce(
+          [](std::unique_ptr<std::string> response_body,
+             ledger::client::LoadURLCallback callback,
+             ledger::mojom::UrlResponse response,
+             scoped_refptr<base::SequencedTaskRunner> post_response_runner) {
+            data_decoder::JsonSanitizer::Sanitize(
+                *response_body,
+                base::BindOnce(&JsonSanitizeCallback, std::move(callback),
+                               std::move(response),
+                               std::move(post_response_runner)));
+          },
+          std::move(response_body), std::move(callback), std::move(response),
+          base::SequencedTaskRunner::GetCurrentDefault()));
```

Comment:
This can go directly in `OnURLLoaderComplete()`. Also, we don't need the comment now.

Reply:
Fixed at https://github.com/brave/brave-core/pull/17354/files#diff-b3817ac19a1c6ad3db2832fc878795ccc8a6f5fc0ae281cdf70d85a3a0a3d324R1100